src/main/drivers/rssi_softpwm.c

```diff
@@ -0,0 +1,235 @@
+
+    case 1:
+        tbits = (tbits << 1) | IORead(rspIO);
+        tstamp[1] = microsISR();
+        ++timerCount;
+        break;
+
+    case 2:
+        tbits = (tbits << 1) | IORead(rspIO);
+        tstamp[2] = microsISR();
+        ++timerCount;
+        EXTIEnable(rspIO, false);
+        rspInProgress = false;
+        break;
+    }
+}
+
+void rssiSoftPwmInit(const rssiSoftPwmConfig_t *configToUse)
+{
+    pRspConfig = configToUse;
+
+    rspIO = IOGetByTag(pRspConfig->ioTag);
+
+    if (!rspIO)
+        return;
+
+    IOInit(rspIO, OWNER_RSSI_SOFTPWM, 0);
+    IOConfigGPIO(rspIO, IOCFG_IPD);
+
+    EXTIHandlerInit(&rsp_extiCallbackRec, rspExtiHandler);
+
+    for (int i = 0 ; i < MEDIAN_TAPCOUNT ; i++)
+        dutyFilterArray[i] = 0;
+
+    dutyFilterPos = 0;
+
+    // Configure EXTI; will trigger on the first rising edge.
+
+    rspInProgress = true;
+    EXTIConfig(rspIO, &rsp_extiCallbackRec, NVIC_PRIO_SOFTPWM_EXTI, EXTI_Trigger_Rising);
+}
+
+//
+// rssiSoftPwmUpdate: Scheduler task
+//
+
+void rssiSoftPwmUpdate(uint32_t currentTime)
+{
+    UNUSED(currentTime);
+
+    if (!rspIO)
+        return;
+
+    int32_t tCycle;
+    int32_t tHigh;
+    int32_t duty;
+
+    uint16_t minDuty = 100;
+
+    ATOMIC_BLOCK(NVIC_PRIO_SOFTPWM_EXTI) {
+        if (rspInProgress) {
+            // Did not finish the current measurement.
+            rspInProgress = false;
+            EXTIEnable(rspIO, false);
+        }
+    }
+
+    if (timerCount != 3) {  // Did not finish, assume no signal
+        duty = 0;
+    } else if (TBITS_IS_NORMAL(tbits)) {
+        tCycle = tstamp[2] - tstamp[0];
+        tHigh = tstamp[1] - tstamp[0];
+
+        // Reject bad measurement
+        // XXX Should find out why this happens (especially the negative case)
+        if (tCycle > 2000 || tCycle < 0) {
+            exceptions = (exceptions & 0xff00) | ((exceptions & 0xff) + 1);
+            goto restart;
+        }
+
+        duty = (tHigh * 100 * 10) / tCycle;
+    } else if (TBITS_MISSED_FALLING_EDGE(tbits)) {
+        // Assume very short pulse
+        duty = 1;
+    } else if (TBITS_MISSED_2ND_RISING_EDGE(tbits)) {
+        // Assume very long pulse
+        duty = 999;
+    } else {
+        // Can't happen!?
+        exceptions = ((exceptions & 0xff00) + 1) | (exceptions & 0xff);
+        goto restart;
+    }
+
+    // Apply median filter to smooth out rapid changes
+
+    dutyFilterArray[dutyFilterPos] = duty;
+    dutyFilterPos = (dutyFilterPos + 1) % MEDIAN_TAPCOUNT;
+    duty = (uint16_t)QUICKMEDIANFILTER(dutyFilterArray);
+
+    if (duty < minDuty)
+        minDuty = duty;
+
+    rspValue = scaleRange(duty, pRspConfig->minFollow ? minDuty : pRspConfig->min, 999, RSSI_OUTPUT_MIN, RSSI_OUTPUT_MAX);
+
+    if (pRspConfig->monitor) {
+        debug[0] = tCycle;
+        debug[1] = tHigh;
+        debug[2] = duty;
+        debug[3] = exceptions;
+    }
+
+restart:
+    // Start a new measurement
+
+    EXTISetTrigger(rspIO, EXTI_Trigger_Rising);
```

Comment:
I may be good idea to clear EXTI bit here (it is still set with interrupt disabled)